### ISSUE TITLE
Add a concept of source to models.

### ISSUE THREAD

**Issue by adamcik** [mopidy/mopidy#1108]
More details in http://goo.gl/TOzY8L

**Comment by jodal** [comment]
Setting tentative milestone to 1.1. We should probably have a last face-to-face discussion about source vs backend in the core API, so we're sure we know what the full usage story of the source fields are.

**Comment by dbrgn** [comment]
I'd like to try this :)

**Comment by adamcik** [comment]
Great, let us know if there isn't enough details in the doc we linked to for this improvement :-)

**Comment by dbrgn** [comment]
So I'd propose the following:

- New readonly string field in the `Track` model called "source", defaults to either `unknown` or `default`
- File and local backends use `local`
- General streams like radio stations use `stream`
- Custom backends set it to values like `spotify` or `soundcloud`

What do you think?

**Comment by dbrgn** [comment]
For many extensions we could probably directly use `Extension.ext_name`. Is that value available e.g. in a `LibraryProvider`?

**Comment by adamcik** [comment]
That was my thinking as well as I was considering having the source just be auto added by core. Where core would look at an attribute on the library / backend and use that, but fall back to the name. But could also work nicely with having the backend set it explicitly if it wants something special, otherwise core sets it to the `ext_name`.

But sounds like you are on the right track.

**Comment by jodal** [comment]
The current state here after the discussion in #1234:

- We want to add a `source` attribute to most/all models.
- We want backends to set the `source` attribute when creating models. No magic happens in Mopidy core.

There's two open questions:

### Exactly which models should have a `source` attribute?

All models, including `Ref`, except `TlTrack` which isn't related to a specific backend. Agree?

### How should backends select what to set the `source` attribute to when creating models?

For music services, this is quite easy. Mopidy-Spotify sets `spotify`, Mopidy-SoundCloud sets `soundcloud`, Mopidy-YouTube sets `youtube`, etc. In general, the extension's `ext_name` is used. If someone creates e.g. Mopidy-Spotify2k17, it should still set the `source` to `spotify`, as the music source is still the Spotify service.

For radio streams, no matter what backend provides the stream, I think the source should be set to `radio`.

For podcasts, no matter what backend is used to find it, I think the source should be set to `podcast`.

Mopidy-Local and Mopidy-File should obviously use the same `source` value, but I'm not sure if this should be `local` or `file`. `file` might be the better choice, as I feel that might cover Mopidy-dLeyna as well? Personally, I don't find it important to separate music from your local disk (which might be mounted using NFS or Samba, and thus be affected by network issues, like dLeyna) and music from your local NAS. @tkem might have thoughts about this.

The decisions here should lead directly to docs explaining the intended usage with examples like the above.

**Comment by tkem** [comment]
@jodal: I don't know. I never really got the motivation behind the `source` attribute, and  I don't see any use case for that in Mopidy-Mobile. At least none that can't be handled by looking at the URI scheme (like handling streams specially in some *rare* cases).
From your proposal, to me it still sounds somewhat like duplicating the information already present in URIs, with extra burden put on extension developers. But I didn't follow the discussion closely, and there may be valid use cases for other clients, of course.

**Comment by jodal** [comment]
Good question... there's been years since we first discussed the `source` attribute.

The motivation is to let extensions show different branding for tracks from different sources without parsing the URIs, which they obviously do today. We want the URIs to be fully opaque and not to be parsed by frontends at all.

Also, the source is intended to a bit more generic than the URI, allowing e.g. two competing extensions or multiple similar extensions to share a source identifier, reducing the number of variations frontends need to care about when branding tracks.

**Comment by tkem** [comment]
And I think this kind of "branding" stuff is generally a bad idea, because as a client developer I don't want to have to keep up with all the new Mopidy extensions that might get popular at some point (and sadly often go looking for a maintainer a short while after) ;-)
So if that's the only proper user story, -1 when wearing my backend developer hat if I have to set the `source` explicitly from my extension.

**Comment by adamcik** [comment]
Other option could be a core call that you pass in URIs and get back the source for all of them. But I would also be ok with just dropping the whole thing. And as for parsing URIs, once we support prefix matches the frontends would have a harder time of it anyway.

**Comment by jodal** [comment]
I think the primary use case is for end users to be quickly able to tell if a track is e.g. from Spotify or from their local music collection. Varying branding per source is just a way to achieve that.

:-1: to having a separate Core method for this. That's another round trip and further complicating the API.

I would say that prefix matching makes `source` attributes even more important, because matching on HTTP URLs like `https://www.youtube.com/...` quickly becomes way more complicated than the current situation with just URI scheme extraction. Thus, `source` attributes should be added to the models before or at the same time as we switch to prefix matching.

It is clearly OK for a backend to not set the `source` attribute, and clients must expect `source` to be missing. However, if we add the `source` attribute to the top extensions, we should quickly be able to get rid of client URI parsing.

@kingosticks @jcass77 AFAIK, Mopidy-Musicbox-Webclient currently parses URIs to add icons next to the tracks. What degree of support for the `source` attribute would you like to see before removing the URI parsing from MMW?

**Comment by tkem** [comment]
@adamcik to the rescue again ;-)

Having a call for retrieving meta information for the backend(s) given a set of URIs might indeed be useful for MoMo too. Maybe we could not only return a "source" for that, but also some user-friendly string for the backend (`Extension.dist_name`, for example, or some name/description set by the backend).

One use case might be https://github.com/tkem/mopidy-mobile/issues/147; I now can retrieve the schemes for playlist providers, but I'd like to present users with something more detailed when selecting a backend...  So if I can, say `getSource({uris: ['m3u:', 'spotify:']})` and get something like

```
{
  'm3u:': {source: 'file' /* or whatever */, name: 'M3U Playlist Provider', ...},
  'spotify:': {source: 'spotify', name: 'Spotify Playlist Provider (read-only)', ... }
}
```
I might get interested, at last ;-) 
   
This "source" information could then be set up once in the backend I guess, similar to `uri_schemes`. No need for attaching `source` for each and every model I return, so I could live with that...

**Comment by jcass77** [comment]
There is a bit of an ongoing debate over at the MMW project as to whether frontends should be offering different user experiences for different backends (e.g. human-friendly names, icons, feature detection), or if we should just straight-up rely on what the Mopidy core provides.

I personally think that frontend developers will always be faced with wanting to add a cool feature that makes their extension easier to use, but that is too specific to have a chance of ever ending up in the core. If you share that view then having a unique URI scheme is all that is required to map various user-configurable options against. I hear what @tkem says about maintenance being and issue though, and maybe just pre-configuring the most popular sources and leaving the rest up to the user might be a good enough compromise?

In terms of features for a 'source' attribute; from a frontend perspective I think this could potentially include:

- UI friendly name
- icon representation
- source service that it references (e.g. ``youtube``)
- whether it is searchable (this is probably something that should be requested from the library?)
- whether it is a type of stream, or if it allows tracks to be looked up and played individually (could be useful in future for deciding if persisting tracklist or playback state between restarts is possible, as well as whether saving tracks from the source to playlists is allowed).

**Comment by tkem** [comment]
I agree with @jcass77 that getting more "meta information" about a model's backend would be desirable. Another use case I can think of would be dissecting search results from multiple backends; if search results would include/provide a way of, let's say, getting a user friendly name (I18N, anybody?) for the backend they came from, this might allow clients to present results in a more user-friendly manner. Still, I think only setting a `source` property to some arbitrary string, which would have to mapped to something meaningful by the client (which wouldn't know about *all* backends), isn't worth the effort.

**Comment by ZenithDK** [comment]
Not having worked on front-ends before I may very well be oblivious to a bunch of problems - but I've personally become quite interested in forking one of the web front-ends that "talks" spotify, and hack it to support Tidal also using https://github.com/mones88/mopidy-tidal
I would need the source here to create support based on what back-end is enabled (maybe only support one of the two, or merge both back-end results into the same UI elements)
